pkg/metrics/controller.go

```diff
@@ -250,7 +259,60 @@ func (c *Controller) recordFleetDeletion(obj interface{}) {
 		return
 	}
 
-	c.recordFleetReplicas(f.Name, f.Namespace, 0, 0, 0, 0, 0)
+	if runtime.FeatureEnabled(runtime.FeatureResetMetricsOnDelete) {
+		if err := c.resyncFleets(); err != nil {
+			// If for some reason resync fails, the entire metric state for fleets
+			// will be reset whenever the next Fleet gets deleted, in which case
+			// we end up back in a healthy state - so we aren't going to actively retry.
+			c.logger.WithError(err).Warn("Could not resync Fleet Metrics")
+		}
+	} else {
+		c.recordFleetReplicas(f.Name, f.Namespace, 0, 0, 0, 0, 0)
+	}
+}
+
+// resyncFleets resets all views associated with a Fleet, and recalculates all totals.
+func (c *Controller) resyncFleets() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	fleets, err := c.fleetLister.List(labels.Everything())
+	if err != nil {
+		return errors.Wrap(err, "could not resync Fleets")
+	}
+
+	fasList, err := c.fasLister.List(labels.Everything())
+	if err != nil {
+		return errors.Wrap(err, "could not resync Fleets")
+	}
+
+	resetViews(fleetViews)
```

Comment:
nit: It looks like `fleetViews` includes and is a superset of `fleetAutoscalerViews` (which it took me a minute to realize). Would it be better to replace this with two lines:

```
resetViews(fleetViews)
resetViews(fleetAutoscalerViews)
```

to make this really obvious without having to find the variable definition in a different file. Then we would change `fleetViews` to no longer append `fleetAutoscalerViews` in controller_metrics.go:43?

Reply:
As per comment below - the FleetAutoscaler views have a fleet label on them - so they are also FleetViews.

I'd argue this is about intent "reset everything to do with fleets", and having the underlying data structures know what the intent should entail.

Reply:
ack.